src/support/allocators/bulk_pool.h

```diff
@@ -0,0 +1,331 @@
+///    [ nextList, T, T, ... T]
+///        v
+///    [ nextList, T, T, ... T]
+///        v
+///      nullptr
+///
+/// m_free_chunks represents a singly linked list of all T's that have been deallocated.
+class Pool
+{
+public:
+    /// Explicitly specify the Pool's chunk size. It will only allocate chunks of this size, returning
+    /// nullptr if a type of different size is specified. If set to 0, the first Allocate() call will
+    /// determine the chunk size of the pool.
+    explicit Pool(size_t chunk_size) noexcept
+        : m_chunk_size(chunk_size)
+    {
+    }
+
+    /// Doesn't specify the chunk size, so it's determined by the first call to Allocate.
+    Pool() = default;
+
+    // Don't allow moving/copying a pool, it's dangerous
+    Pool(Pool&&) = delete;
+    Pool& operator=(Pool&&) = delete;
+    Pool(const Pool&) = delete;
+    Pool& operator=(const Pool&) = delete;
+
+    /// Don't allow allocation for types that are smaller than a void*. This does not make sense
+    /// because we need to fit a pointer into the memory.
+    template <typename T, typename std::enable_if<sizeof(T) < sizeof(void*), int>::type = 0>
+    T* Allocate() noexcept
+    {
+        return nullptr;
+    }
+
+    /// As with Allocate(), don't allow for types that are smaller than a void*.
+    template <typename T, typename std::enable_if<sizeof(T) < sizeof(void*), int>::type = 0>
+    bool Deallocate() noexcept
+    {
+        return false;
+    }
+
+    /// Tries to allocate one T. If allocation is not possible, returns nullptr and the callee
+    /// has to do something else to get memory. First caller decides the size of the pool's data.
+    template <typename T, typename std::enable_if<sizeof(T) >= sizeof(void*), int>::type = 0>
+    T* Allocate()
+    {
+        if (m_chunk_size == 0) {
+            // allocator not yet used, so this type determines the size.
+            m_chunk_size = sizeof(T);
+        } else if (m_chunk_size != sizeof(T)) {
+            // allocator's size does not match sizeof(T), don't allocate.
+            return nullptr;
+        }
+
+        // Make sure we have memory available
+        if (m_free_chunks == nullptr) {
+            AllocateAndCreateFreelist();
+        }
+
+        // pop one element from the linked list, returning previous head
+        auto old_head = m_free_chunks;
+        m_free_chunks = *reinterpret_cast<void**>(old_head);
+        return reinterpret_cast<T*>(old_head);
+    }
+
+    /// Puts p back into the freelist, if it was the correct size. Only allowed with objects
+    /// that were allocated with this pool!
+    template <typename T, typename std::enable_if<sizeof(T) >= sizeof(void*), int>::type = 0>
+    bool Deallocate(T* p) noexcept
+    {
+        if (m_chunk_size != sizeof(T)) {
+            // allocation didn't happen with this allocator
+            return false;
+        }
+
+        // put it into the linked list
+        *reinterpret_cast<void**>(p) = m_free_chunks;
```

Comment:
hm, I could add a 
```cpp
struct Node {
    Node* next;
};
```
And then use `Node*` instead of `void*`. Then instead of 
```cpp
// put it into the linked list
*reinterpret_cast<void**>(p) = m_free_chunks;
m_free_chunks = p;
```
I could write 
```cpp
auto n = reinterpret_cast<Node*>(p);
n->next = m_free_chunks;
m_free_chunks = n;
```
I guess the code is a bit better readable with that, but I honestly don't know if it's more defined behavior :thinking:

Reply:
I got rid of the `void**` casts in in https://github.com/bitcoin/bitcoin/pull/16801/commits/52ee5735304cb8be342e264a19274afc0f422fa8